README.md
# Daisy

Rust `no_std`, `embedded_hal` board support package for the Electro-Smith Daisy
platform.

* [Documentation](https://zlosynth.com/daisy)
* [Crate (crates.io)](https://crates.io/crates/daisy)
* [Repository (github.com)](https://github.com/zlosynth/daisy)

# Supported boards

Currently this library supports following boards:

* [Daisy Seed](https://www.electro-smith.com/daisy/daisy) (codec AK4556), `seed`
* [Daisy Seed 1.1](https://www.electro-smith.com/daisy/daisy) (codec WM8731), `seed_1_1`
* [Daisy Seed 1.2](https://www.electro-smith.com/daisy/daisy) (codec PCM3060), `seed_1_2`
* [Daisy Patch SM](https://www.electro-smith.com/daisy/patch-sm) (codec PCM3060), `patch_sm`

Select the board by using its respective feature.

# Flashing an example

``` sh
make flash WHAT=blinky BOARD=seed_1_1
```

# Sampling rate

By default, the audio sampling rate is set to 48 kHz. This can be increased to
96 kHz by enabling the `sampling_rate_96khz` feature.

# API stability

I am still trying to figure out a good API for the project. Expect it to change.
To mitigate breakage of your code on library update, use macros defined under
`board.rs` to initialize resources whenever possible.

# HAL compatibility

This library is closely tied to [stm32h7xx-hal](https://github.com/stm32-rs/stm32h7xx-hal/).
Make sure to use compatible versions in your `Cargo.toml`.

| **Daisy**   | **HAL** |
|-------------|---------|
| `0.8`       | `0.14`  |
| `0.2`-`0.7` | `0.12`  |
| `0.1`       | `0.11`  |

# Razor

* This library aims to abstract all peripherals that are available on Daisy.
* This library will **not** abstract anything that is not on Daisy, e.g. pots,
  SD cards, MIDI, ...

# License

This library is distributed under the terms of the MIT license. See
[LICENSE](LICENSE) for details.

This project was forked from
[antoinevg/daisy_bsp](https://github.com/antoinevg/daisy_bsp).
Kudos to Antoine for making his library public.

# Changelog

Read the [CHANGELOG.md](CHANGELOG.md) to learn about changes introduced in each
release.

# Versioning

See [VERSIONING.md](VERSIONING.md) to find detailed information about versioning
of the project.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[VERSIONING.md](VERSIONING.md) for more information about versioning and
backwards compatibility.

## Unreleased

* Allow audio sampling rate of 96 kHz.
* Introduce support for Daisy Seed 1.2.

## 0.8.0

* **Breaking** Remove logging abstractions from the crate.
* **Breaking** The crate no longer re-exports `embedded_hal`.
* **Breaking** Remove `Led` trait.
* Bump STM32H7 HAL to version 0.14.
* Improve style and documentation of examples.
* Introduce RTIC example of audio processing.
* Introduce example of SD card interaction.

## 0.7.3

* Fix occasional stuckage of flash initialization.

## 0.7.2

* Set correct ADC/DAC format on Seed 1.1 and Patch SM to solve conversion issue.

## 0.7.1

* Fix audio input processing for Seed 1.1 and Patch SM.

## 0.7.0

* Breaking: Remove `size` parameter from SDRAM initialization.
* Allow utilization of dcache.
* Publish documentation under [https://zlosynth.com/daisy](https://zlosynth.com/daisy).

## 0.6.0

* Introduce an interface to access SDRAM.
* Breaking: Audio processing callback is now passed to
  `handle_interrupt_dma1_str1` instead of `spawn`.
* Breaking: Rename `FMCPins` to `FlashPins`.

## 0.5.0

* Add an example of OLED display configuration.
* Configure clock to support displays connected via SPI.

## 0.4.5

* Bump cortex-m-rt dependency to 0.7.

## 0.4.4

* Implicitly use peripheral clock for ADC.
* Fix ITM module.

## 0.4.3

* Fix mapping of pins B5, B6 and D10.

## 0.4.2

* Fix ADC pins C6-C9 mapping.

## 0.4.1

* Fix noise in Patch SM audio interface by switching to I2S.

## 0.4.0

* Fix macros so they respect the feature set on the depedency.
* Introduce a macro to instantiate flash abstraction.
* Update to stm32h7xx HAL 0.12.1.

## 0.3.0

* Introduce support for Daisy Patch SM, with PCM3060 codec.
* Move general purpose pins under a sub-structure. e.g. `SEED_PIN_1` to
  `GPIO.PIN_1`.
* Rename `seed_1_0` feature to `seed`.
* Remove unneeded `cty` dependency.

## 0.2.0

* Update to stm32h7xx HAL 0.12.
* Introduce support for Daisy Seed 1.1, with WM8731 codec.

## 0.1.0

* Initial fork from [antoinevg/daisy_bsp](https://github.com/antoinevg/daisy_bsp).
* Added flash memory access interface.
* Define razor guiding what goes into the library and what not.
* Introduce CI.
* Add SRAM to memory layout.


examples/adc.rs
//! Example of how to directly access the ADC peripheral on the Daisy.

#![no_main]
#![no_std]

use cortex_m::asm;
use cortex_m_rt::entry;
use panic_semihosting as _;

use hal::adc;
use hal::delay::Delay;
use hal::prelude::*;
use stm32h7xx_hal as hal;

#[entry]
fn main() -> ! {
    // Get core and device peripherals, and the board abstraction.
    let cp = cortex_m::Peripherals::take().unwrap();
    let dp = daisy::pac::Peripherals::take().unwrap();
    let board = daisy::Board::take().unwrap();

    // Configure board's peripherals.
    let ccdr = daisy::board_freeze_clocks!(board, dp);
    let pins = daisy::board_split_gpios!(board, ccdr, dp);

    // Configure ADC.
    let mut delay = Delay::new(cp.SYST, ccdr.clocks);
    let mut adc1 = adc::Adc::adc1(
        dp.ADC1,
        4.MHz(),
        &mut delay,
        ccdr.peripheral.ADC12,
        &ccdr.clocks,
    )
    .enable();
    adc1.set_resolution(adc::Resolution::SixteenBit);

    // Select a pin that will be used for ADC, depending on the board.
    #[cfg(any(feature = "seed", feature = "seed_1_1", feature = "seed_1_2"))]
    let mut adc1_channel = pins.GPIO.PIN_21.into_analog();
    #[cfg(feature = "patch_sm")]
    let mut adc1_channel = pins.GPIO.PIN_C2.into_analog(); // CV_4

    // Get a handle on the on-board LED to later use as an indicator.
    let mut led_user = pins.LED_USER.into_push_pull_output();

    // Divide the processor clock with 2^16 (for 16 bit ADC resolution configured above).
    // Scaling by the result will produce zero delay for the lowest ADC value and
    // 1 second for the highest.
    let scale_factor = ccdr.clocks.sys_ck().to_Hz() as f32 / 65_535.0;

    loop {
        // Read value from the ADC pin.
        let pot: u32 = adc1.read(&mut adc1_channel).unwrap();

        // Scale the value to a delay counted in CPU ticks.
        let ticks = (pot as f32 * scale_factor) as u32;

        // Blink the LED with interval depending on the read value.
        led_user.set_high();
        asm::delay(ticks);
        led_user.set_low();
        asm::delay(ticks);
    }
}


examples/oled.rs
//! Example of interfacing with an external OLED display.

#![no_main]
#![no_std]

use cortex_m_rt::entry;
use panic_semihosting as _;

use hal::prelude::*;
use hal::{delay::DelayFromCountDownTimer, spi};
use stm32h7xx_hal as hal;

use embedded_graphics::{
    pixelcolor::BinaryColor,
    prelude::*,
    primitives::{Circle, PrimitiveStyle, Rectangle},
};
use fugit::RateExtU32;
use ssd1306::{prelude::*, Ssd1306};

#[entry]
fn main() -> ! {
    // Get device peripherals and the board abstraction.
    let board = daisy::Board::take().unwrap();
    let dp = daisy::pac::Peripherals::take().unwrap();

    // Configure board's peripherals.
    let ccdr = daisy::board_freeze_clocks!(board, dp);
    let pins = daisy::board_split_gpios!(board, ccdr, dp);

    // Configure the display.
    //
    // This example uses 128x64 display connected via 4-wire SPI. The pin
    // mapping is:
    //
    // | Interface role | Display label | Seed pin | Patch SM pin |
    // |----------------|---------------|----------|--------------|
    // | SCK            | CLK           | 9        | D10          |
    // | CS             | CS            | 8        | D1           |
    // | MOSI           | DIN           | 11       | D9           |
    // | RST            | RES           | 37       | A9           |
    // | DC             | D/C           | 10       | D8           |
    let mut display = {
        // Select the pins depending on the used board.
        #[cfg(any(feature = "seed", feature = "seed_1_1", feature = "seed_1_2"))]
        let pins = (
            pins.GPIO.PIN_8.into_alternate(),
            pins.GPIO.PIN_7.into_push_pull_output(),
            pins.GPIO.PIN_10.into_alternate(),
            pins.GPIO.PIN_30.into_push_pull_output(),
            pins.GPIO.PIN_9.into_push_pull_output(),
        );
        #[cfg(feature = "patch_sm")]
        let pins = (
            pins.GPIO.PIN_D10.into_alternate(),
            pins.GPIO.PIN_D1.into_push_pull_output(),
            pins.GPIO.PIN_D9.into_alternate(),
            pins.GPIO.PIN_A9.into_push_pull_output(),
            pins.GPIO.PIN_D8.into_push_pull_output(),
        );
        let (sck, cs, mosi, mut rst, dc) = pins;

        // Initialize SPI, selecting the right peripheral depending on the used
        // board.
        #[cfg(any(feature = "seed", feature = "seed_1_1", feature = "seed_1_2"))]
        let spi = dp.SPI1.spi(
            (sck, spi::NoMiso, mosi),
            spi::MODE_0,
            3.MHz(),
            ccdr.peripheral.SPI1,
            &ccdr.clocks,
        );
        #[cfg(feature = "patch_sm")]
        let spi = dp.SPI2.spi(
            (sck, spi::NoMiso, mosi),
            spi::MODE_0,
            3.MHz(),
            ccdr.peripheral.SPI2,
            &ccdr.clocks,
        );

        // Configure the display interface and its driver.
        let interface = display_interface_spi::SPIInterface::new(spi, dc, cs);
        let mut display = Ssd1306::new(interface, DisplaySize128x64, DisplayRotation::Rotate0)
            .into_buffered_graphics_mode();
        let mut delay = DelayFromCountDownTimer::new(dp.TIM2.timer(
            100.Hz(),
            ccdr.peripheral.TIM2,
            &ccdr.clocks,
        ));
        display.reset(&mut rst, &mut delay).unwrap();
        display.init().unwrap();

        display
    };

    // Draw basic shapes on the display.
    let one_second = ccdr.clocks.sys_ck().to_Hz();
    loop {
        Rectangle::new(Point::new(0, 0), Size::new(127, 63))
            .into_styled(PrimitiveStyle::with_fill(BinaryColor::Off))
            .draw(&mut display)
            .unwrap();
        Circle::new(Point::new(39, 7), 51)
            .into_styled(PrimitiveStyle::with_fill(BinaryColor::On))
            .draw(&mut display)
            .unwrap();
        display.flush().unwrap();
        cortex_m::asm::delay(2 * one_second);

        Rectangle::new(Point::new(0, 0), Size::new(127, 63))
            .into_styled(PrimitiveStyle::with_fill(BinaryColor::On))
            .draw(&mut display)
            .unwrap();
        Circle::new(Point::new(39, 7), 51)
            .into_styled(PrimitiveStyle::with_fill(BinaryColor::Off))
            .draw(&mut display)
            .unwrap();
        display.flush().unwrap();
        cortex_m::asm::delay(2 * one_second);
    }
}


examples/sdmmc.rs
//! Example of interaction with an SD card connected to the module's SDMMC interface.

#![no_main]
#![no_std]

use cortex_m::asm;
use cortex_m_rt::entry;
use panic_semihosting as _;

use hal::gpio::Speed;
use hal::sdmmc::{SdCard, Sdmmc};
use hal::{pac, prelude::*};
use stm32h7xx_hal as hal;

#[entry]
fn main() -> ! {
    // Get core and device peripherals, and the board abstraction.
    let board = daisy::Board::take().unwrap();
    let dp = pac::Peripherals::take().unwrap();

    // Configure board's peripherals.
    let ccdr = daisy::board_freeze_clocks!(board, dp);
    let pins = daisy::board_split_gpios!(board, ccdr, dp);
    let mut led_user = daisy::board_split_leds!(pins).USER;
    let one_second = ccdr.clocks.sys_ck().to_Hz();

    // Select pins connected to the SD card depending on the used board.
    #[cfg(any(feature = "seed", feature = "seed_1_1", feature = "seed_1_2"))]
    let (clk, cmd, d0, d1, d2, d3) = (
        pins.GPIO.PIN_6,
        pins.GPIO.PIN_5,
        pins.GPIO.PIN_4,
        pins.GPIO.PIN_3,
        pins.GPIO.PIN_2,
        pins.GPIO.PIN_1,
    );
    #[cfg(feature = "patch_sm")]
    let (clk, cmd, d0, d1, d2, d3) = (
        pins.GPIO.PIN_D6,
        pins.GPIO.PIN_D7,
        pins.GPIO.PIN_D5,
        pins.GPIO.PIN_D4,
        pins.GPIO.PIN_D3,
        pins.GPIO.PIN_D2,
    );

    // Configure the pins.
    let clk = clk
        .into_alternate::<12>()
        .internal_pull_up(false)
        .speed(Speed::VeryHigh);
    let clk = clk
        .into_alternate()
        .internal_pull_up(false)
        .speed(Speed::VeryHigh);
    let cmd = cmd
        .into_alternate()
        .internal_pull_up(true)
        .speed(Speed::VeryHigh);
    let d0 = d0
        .into_alternate()
        .internal_pull_up(true)
        .speed(Speed::VeryHigh);
    let d1 = d1
        .into_alternate()
        .internal_pull_up(true)
        .speed(Speed::VeryHigh);
    let d2 = d2
        .into_alternate()
        .internal_pull_up(true)
        .speed(Speed::VeryHigh);
    let d3 = d3
        .into_alternate()
        .internal_pull_up(true)
        .speed(Speed::VeryHigh);

    // Initialize abstraction of the SDMMC interface.
    let mut sdmmc: Sdmmc<_, SdCard> = dp.SDMMC1.sdmmc(
        (clk, cmd, d0, d1, d2, d3),
        ccdr.peripheral.SDMMC1,
        &ccdr.clocks,
    );

    // This can be increased up to 50MHz. We choose a lower frequency here so
    // that it should work even with flying leads connected to a SD card breakout.
    let bus_frequency = 2.MHz();

    // Try to connect to the SD card. Blink rapidly to signal waiting.
    while sdmmc.init(bus_frequency).is_err() {
        led_user.toggle();
        asm::delay(one_second / 8);
    }

    // Write to the card, read back, test that the values match.
    let mut buffer = [0x34; 512];
    sdmmc.write_block(0, &buffer).unwrap();
    sdmmc.read_block(0, &mut buffer).unwrap();
    for byte in buffer.iter() {
        assert_eq!(*byte, 0x34);
    }

    // Keep blinking to block main and shows signs of life and to show that
    // the test above passed.
    let one_second = ccdr.clocks.sys_ck().to_Hz();
    loop {
        led_user.toggle();
        asm::delay(one_second);
    }
}


src/audio/interface.rs
use core::num::Wrapping;

use super::codec::{Codec, Pins as CodecPins};
use super::transfer::{Channel, Config as TransferConfig, Sai1Pins, State, Sync, Transfer};
use super::{BLOCK_LENGTH, DMA_BUFFER_LENGTH, FS, HALF_DMA_BUFFER_LENGTH};
use crate::hal;
use hal::pac::{CorePeripherals, CPUID};
use hal::time;

#[link_section = ".sram1_bss"]
static mut TX_BUFFER: [u32; DMA_BUFFER_LENGTH] = [0; DMA_BUFFER_LENGTH];
#[link_section = ".sram1_bss"]
static mut RX_BUFFER: [u32; DMA_BUFFER_LENGTH] = [0; DMA_BUFFER_LENGTH];

pub type Frame = (f32, f32);
pub type Block = [Frame; BLOCK_LENGTH];

#[derive(Debug)]
pub enum Error {
    I2c,
    Dma,
}

pub struct Interface {
    pub fs: time::Hertz,
    codec: Codec,
    transfer: Transfer,
}

impl Interface {
    pub fn init(
        clocks: &hal::rcc::CoreClocks,
        sai1_rec: hal::rcc::rec::Sai1,
        sai1_pins: Sai1Pins,
        codec_pins: CodecPins,
        #[allow(unused_variables)] // i2c2 is not used on Seed 1.0
        i2c2_rec: hal::rcc::rec::I2c2,
        dma1_rec: hal::rcc::rec::Dma1,
    ) -> Result<Interface, Error> {
        #[cfg(any(feature = "seed", feature = "seed_1_2"))]
        let codec = Codec::init(codec_pins);
        #[cfg(any(feature = "seed_1_1", feature = "patch_sm"))]
        let codec = Codec::init(clocks, i2c2_rec, codec_pins);

        #[cfg(feature = "seed")]
        let transfer_config = TransferConfig {
            tx_channel: Channel::A,
            rx_channel: Channel::B,
            tx_sync: Sync::Master,
            rx_sync: Sync::Slave,
        };
        #[cfg(any(feature = "seed_1_1", feature = "patch_sm"))]
        let transfer_config = TransferConfig {
            tx_channel: Channel::B,
            rx_channel: Channel::A,
            tx_sync: Sync::Slave,
            rx_sync: Sync::Master,
        };
        #[cfg(feature = "seed_1_2")]
        let transfer_config = TransferConfig {
            tx_channel: Channel::B,
            rx_channel: Channel::A,
            tx_sync: Sync::Master,
            rx_sync: Sync::Slave,
        };

        let transfer = Transfer::init(
            clocks,
            sai1_rec,
            sai1_pins,
            dma1_rec,
            unsafe { &mut TX_BUFFER },
            unsafe { &mut RX_BUFFER },
            transfer_config,
        );

        // Verifying safety requirements and recommendations of
        // invalidate_dcache_by_slice and clean_dcache_by_slice.
        validate_slice_against_cache_line(unsafe { &TX_BUFFER });
        validate_slice_against_cache_line(unsafe { &RX_BUFFER });

        Ok(Self {
            fs: FS,
            codec,
            transfer,
        })
    }

    /// Start audio streaming.
    pub fn spawn(mut self) -> Result<Self, Error> {
        self.codec.start();
        self.transfer.start();
        Ok(self)
    }

    pub fn handle_interrupt_dma1_str1(
        &mut self,
        mut callback: impl FnMut(&mut [(f32, f32); BLOCK_LENGTH]),
    ) -> Result<(), Error> {
        let skip = match self.transfer.examine_interrupt() {
            Ok(State::HalfSent) => (0, HALF_DMA_BUFFER_LENGTH),
            Ok(State::FullSent) => (HALF_DMA_BUFFER_LENGTH, 0),
            _ => return Err(Error::Dma),
        };

        let mut block: Block = [(0.0, 0.0); BLOCK_LENGTH];

        // Force dcache to get populated from memory.
        // Safety: RX buffer is accessed only through this function, without any
        // concurrency. The init function verifies that the buffer has the
        // correct size and alignment.
        unsafe {
            CorePeripherals::steal()
                .SCB
                .invalidate_dcache_by_slice(&mut RX_BUFFER);
        }

        // Convert and copy received audio to callback buffer.
        let mut dma_index: usize = 0;
        let mut block_index: usize = 0;
        while dma_index < HALF_DMA_BUFFER_LENGTH {
            let rx0: usize = dma_index + skip.1;
            let rx1: usize = rx0 + 1;

            let rx_y0 = unsafe { RX_BUFFER[rx0] };
            let rx_y1 = unsafe { RX_BUFFER[rx1] };

            let y0 = u24_to_f32(rx_y0);
            let y1 = u24_to_f32(rx_y1);
            block[block_index] = (y1, y0);

            dma_index += 2;
            block_index += 1;
        }

        // Invoke user-supplied callback.
        callback(&mut block);

        // Convert and copy callback buffer to output audio buffer.
        let mut dma_index: usize = 0;
        let mut block_index: usize = 0;
        while dma_index < HALF_DMA_BUFFER_LENGTH {
            let tx0: usize = dma_index + skip.0;
            let tx1: usize = tx0 + 1;

            let (y0, y1) = block[block_index];
            unsafe { TX_BUFFER[tx0] = f32_to_u24(y0) };
            unsafe { TX_BUFFER[tx1] = f32_to_u24(y1) };

            dma_index += 2;
            block_index += 1;
        }

        // Force dcache to get flushed into memory.
        // Safety: TX buffer is accessed only through this function, without any
        // concurrency. The init function verifies that the buffer has the
        // correct size and alignment.
        unsafe {
            CorePeripherals::steal()
                .SCB
                .clean_dcache_by_slice(&TX_BUFFER);
        }

        Ok(())
    }
}

/// Verifying safety requirements and recommendations of
/// invalidate_dcache_by_slice and clean_dcache_by_slice. The slice must be
/// aligned with cache lines.
fn validate_slice_against_cache_line<T>(slice: &[T]) {
    let addr = slice.as_ptr() as usize;
    let size = slice.len() * core::mem::size_of_val(slice);
    let dminline = CPUID::cache_dminline();
    let line_size = (1 << dminline) * 4;
    assert!((addr & (line_size - 1)) == 0);
    assert!((size & (line_size - 1)) == 0);
}

/// Convert audio data from u24 to f32.
#[inline(always)]
fn u24_to_f32(y: u32) -> f32 {
    let y = (Wrapping(y) + Wrapping(0x0080_0000)).0 & 0x00FF_FFFF; // convert to i32
    (y as f32 / 8_388_608.0) - 1.0 // (2^24) / 2
}

/// Convert audio data from f32 to u24.
#[inline(always)]
fn f32_to_u24(x: f32) -> u32 {
    let x = x * 8_388_607.0;
    let x = x.clamp(-8_388_608.0, 8_388_607.0);
    (x as i32) as u32
}


src/lib.rs
#![no_std]

//! Rust `no_std`, `embedded_hal` board support package for the Electro-Smith Daisy platform.
//!
//! * [Documentation](https://zlosynth.com/daisy)
//! * [Crate (crates.io)](https://crates.io/crates/daisy)
//! * [Repository (github.com)](https://github.com/zlosynth/daisy)
//!
//! # Supported boards
//!
//! Currently this library supports following boards:
//!
//! * [Daisy Seed](https://www.electro-smith.com/daisy/daisy) (codec AK4556), `seed`
//! * [Daisy Seed 1.1](https://www.electro-smith.com/daisy/daisy) (codec WM8731), `seed_1_1`
//! * [Daisy Seed 1.2](https://www.electro-smith.com/daisy/daisy) (codec PCM3060), `seed_1_2`
//! * [Daisy Patch SM](https://www.electro-smith.com/daisy/patch-sm) (codec PCM3060), `patch_sm`
//!
//! Select the board by using its respective feature.
//!
//! # Sampling rate
//!
//! By default, the audio sampling rate is set to 48 kHz. This can be increased to
//! 96 kHz by enabling the `sampling_rate_96khz` feature.
//!
//!
//! # API stability
//!
//! I am still trying to figure out a good API for the project. Expect it to change.
//! To mitigate breakage of your code on library update, use macros defined under
//! `board.rs` to initialize resources whenever possible.
//!
//! # HAL compatibility
//!
//! This library is closely tied to [stm32h7xx-hal](https://github.com/stm32-rs/stm32h7xx-hal/).
//! Make sure to use compatible versions in your `Cargo.toml`.
//!
//! | **Daisy**   | **HAL** |
//! |-------------|---------|
//! | `0.8`       | `0.14`  |
//! | `0.2`-`0.7` | `0.12`  |
//! | `0.1`       | `0.11`  |
//!
//! # Usage
//!
//! See the [examples/](https://github.com/zlosynth/daisy/tree/main/examples)
//! directory to find usage examples:
//!
//! * [ADC](https://github.com/zlosynth/daisy/blob/main/examples/adc.rs)
//! * [Audio](https://github.com/zlosynth/daisy/blob/main/examples/audio.rs)
//! * [Audio with RTIC](https://github.com/zlosynth/daisy/blob/main/examples/audio_rtic.rs)
//! * [Blinky](https://github.com/zlosynth/daisy/blob/main/examples/blinky.rs)
//! * [Blinky with RTIC](https://github.com/zlosynth/daisy/blob/main/examples/blinky_rtic.rs)
//! * [Flash storage](https://github.com/zlosynth/daisy/blob/main/examples/flash.rs)
//! * [OLED display](https://github.com/zlosynth/daisy/blob/main/examples/oled.rs)
//! * [SDRAM memory](https://github.com/zlosynth/daisy/blob/main/examples/sdram.rs)
//! * [SD card](https://github.com/zlosynth/daisy/blob/main/examples/sdmmc.rs)
//!
//! ``` sh
//! make flash WHAT=blinky BOARD=seed_1_1
//! ```

#[cfg(all(
    feature = "seed_1_2",
    any(feature = "seed_1_1", feature = "seed", feature = "patch_sm")
))]
compile_error!("only a single target board must be selected");

#[cfg(all(
    feature = "seed_1_1",
    any(feature = "seed_1_2", feature = "seed", feature = "patch_sm")
))]
compile_error!("only a single target board must be selected");

#[cfg(all(
    feature = "seed",
    any(feature = "seed_1_2", feature = "seed_1_1", feature = "patch_sm")
))]
compile_error!("only a single target board must be selected");

#[cfg(all(
    feature = "patch_sm",
    any(feature = "seed_1_2", feature = "seed_1_1", feature = "seed")
))]
compile_error!("only a single target board must be selected");

#[cfg(not(any(
    feature = "seed_1_2",
    feature = "seed_1_1",
    feature = "seed",
    feature = "patch_sm"
)))]
compile_error!(
    "target board must be selected using a feature: \"seed_1_2\" | \"seed_1_1\" | \"seed\" | \"patch_sm\""
);

pub mod audio;
pub mod board;
pub mod clocks;
pub mod flash;
pub mod led;
pub mod pins;
pub mod sdram;

pub use board::Board;
pub use hal::pac;
// Re-exported so it can be used from daisy macros.
pub use stm32h7xx_hal as hal;
